[INTERNAL] Enhance ESSupport.md with further details (#737)
SAP/ui5-tooling: docs/pages/CodeAnalysis.md

The bundling implemented by UI5 Tooling can either follow only eager dependencies (resolve:`true`) or additionally conditional dependencies (resolveConditional). For more information, see [Custom Bundling](https://sap.github.io/ui5-tooling/v3/pages/Configuration/#custom-bundling).

When a dependency in one of the mentioned APIs is not a "simple literal" but an expression, the corresponding module is marked as "having dynamic dependencies". This marker is currently not further evaluated by the tooling.

### Component Analyzer

The [Component Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/ComponentAnalyzer.js) analyzes JavaScript files named `Component.js` to collect dependency information by searching for a `manifest.json` in the same folder. If one is found, the `sap.ui5` section is evaluated in the following way:

- Any library dependency is added as a dependency to the `library.js` module of that library. If the library dependency is modeled as 'lazy', the
 module dependency will be added as 'conditional',

> - any component dependency is added as a dependency to a file named `Component.js`. If the dependency is modeled as 'lazy', the module dependency will be added as 'conditional',

Author: @KlattG now some words starting with uppercase in lists and some in lowercase. Could you please have another look on them?

Author: Ah got it, you have structured the list as a sentence.

Reviewer: That is actually fine, because the list consists of full sentences that continue over several list points. In this case the usual sentence capitalization should be used.

Author: Just learned something new :-) Thanks @KlattG

Reviewer: I just had a look into the SAP Style Guide - they recommend to begin the first word of each list item with a letter in the same case. Since we have full sentences here, I'd propose to change all first words to uppercase in this list.

> - Any library dependency is added as a dependency to the `library.js` module of that library. If the library dependency is modeled as 'lazy', the
>  module dependency will be added as 'conditional',
> - any component dependency is added as a dependency to a file named `Component.js`. If the dependency is modeled as 'lazy', the module dependency will be added as 'conditional',
> - for each UI5 model for which a type is configured, a module dependency to that type is added,
> - for each route that contains a view name, a module dependency to that view will be added.

Reviewer: ```suggestion
- Any library dependency is added as a dependency to the `library.js` module of that library. If the library dependency is modeled as 'lazy', the
 module dependency will be added as 'conditional'.
- Any component dependency is added as a dependency to a file named `Component.js`. If the dependency is modeled as 'lazy', the module dependency will be added as 'conditional'.
- For each UI5 model for which a type is configured, a module dependency to that type is added.
- For each route that contains a view name, a module dependency to that view is added.
```

### Fiori Elements Analyzer

> The [Fiori Elements Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/FioriElementsAnalyzer.js) analyzes a FioriElements app and its underlying template components to collect dependency information. It searches for a `manifest.json` in the same folder. If one is found the `sap.fe` section will be evaluated in the following way:

Reviewer: "It searches for a `manifest.json` in the same folder."

Same folder as what?

Author: @KlattG What about "It searches for a `manifest.json` located in the same folder as the `Component.js`"

Reviewer: That sounds fine, thanks!

Author: Done

> - for each entity set in the `entitySets` object, each sub-entry is checked for a `default.template` property

Reviewer: ```suggestion
- For each entity set in the `entitySets` object, each sub-entry is checked for a `default.template` property,
```

> - if found, the containing string is interpreted as the short name of a template component in the `sap.fe.templates` library

Reviewer: ```suggestion
- if one is found, the containing string is interpreted as the short name of a template component in the `sap.fe.templates` library,
```

> - For each entity set in the `entitySets` object, each sub-entry is checked for a `default.template` property,
> - if one is found, the containing string is interpreted as the short name of a template component in the `sap.fe.templates` library,
> - a dependency to that template component is added to the analyzed app.

Reviewer: ```suggestion
- For each entity set in the `entitySets` object, each sub-entry is checked for a `default.template` property.
- If one is found, the containing string is interpreted as the short name of a template component in the `sap.fe.templates` library.
- A dependency to that template component is added to the analyzed app.
```

> - a dependency to that template component is added to the analyzed app

Reviewer: ```suggestion
- a dependency to that template component is added to the analyzed app.
```

For a full analysis, "routing" also should be taken into account. Only when a sub-entry of the entity set

> is referenced by a route, then the template for that entry will be used. Routes thereby could form entry points.

Reviewer: ```suggestion
is referenced by a route, the template for that entry will be used. Routes thereby could form entry points.
```

Author: done

```json
{
    "sap.fe" : {
        "entitySets" : {
            "C_AIVS_MDBU_ArtistTP" : {
                "feed": {
                    "default": {
                        "template": "ListReport"
                    }
                },
                "entry" : {
                    "default" : {
                        "outbound" : "musicV2Display"
                    }
                }
            }
        },
        "routing" : {
            "routes" :{
                "ArtistList": {
                    "target": "C_AIVS_MDBU_ArtistTP/feed"
                }
            }
        }
    }
}
```

The template component is analyzed in the following way:

> - precondition: template component class is defined in an AMD-style module, using define or sap.ui.define

Reviewer: ```suggestion
- precondition: the template component class is defined in an AMD-style module, using define or sap.ui.define
```

> - precondition: the module 'sap/fe/core/TemplateAssembler' is imported

Reviewer 2: --> backticks for sap/fe/core/TemplateAssembler

Author: Done

> - precondition: a call to TemplateAssembler.getTemplateComponent is used to define the component class

Reviewer 2: --> wrap TemplateAssembler.getTemplateComponent in backticks

Author: done

> - precondition: that call is used in a top level return statement of the factory function

Reviewer: ```suggestion
- precondition: that call is used in a top-level return statement of the factory function
```

- precondition: necessary parameters to that call are given as an object literal (no further coding)

> - precondition: the settings define a managed property 'metadata.properties.templateName' with a

Reviewer 2: --> backticks for 'metadata.properties.templateName'

Author: done

> - precondition: the settings define a managed property property 'metadata.properties.templateName' with a

Reviewer: ```suggestion
- precondition: the settings define a managed property 'metadata.properties.templateName' with a
```

                defaultValue of type string
The default value of the property represents the template view of the template component.

> The manifest of the template app could in theory specify an alternative template as setting.templateName,

Reviewer 2: --> backticks for setting.templateName

> The manifest of the template app in theory could specify an alternative template as setting.templateName,

Reviewer: ```suggestion
The manifest of the template app could in theory specify an alternative template as setting.templateName,
```

> but as of June 2017, this possibility is currently not used.

Reviewer 2: --> "but this possibility is currently not used."

Author: done

### Smart Template Analyzer

> The [Smart Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/SmartTemplateAnalyzer.js) analyzes a Smart Template app and its underlying template components to collect dependency information. It searches for a `manifest.json` located in the same folder as the `Component.js`. If it is found and if it is a valid JSON, an "sap.ui.generic.app" section is searched and evaluated in the following way:

Reviewer 2: --> "if it is valid JSON"

Reviewer 2: backticks for "sap.ui.generic.app"

Author: both done

> The [Smart Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/SmartTemplateAnalyzer.js) analyzes a Smart Template app and its underlying template components to collect dependency information. It searches for a `manifest.json` located in the same folder as the `Component.js`. If it is found and if it is a valid JSON, an "sap.ui.generic.app" section is searched and evaluated in the following way

Reviewer: ```suggestion
The [Smart Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/SmartTemplateAnalyzer.js) analyzes a Smart Template app and its underlying template components to collect dependency information. It searches for a `manifest.json` located in the same folder as the `Component.js`. If it is found and if it is a valid JSON, an "sap.ui.generic.app" section is searched and evaluated in the following way:
```

> The [Smart Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/SmartTemplateAnalyzer.js) analyzes a Smart Template app and its underlying template components to collect dependency information. It searches for a `manifest.json` in the same folder. If it is found and if it is a valid JSON, an "sap.ui.generic.app" section is searched and evaluated in the following way

Reviewer: It searches for a `manifest.json` in the same folder.

see my comment above

Reviewer: It searches for a `manifest.json` in the same folder as the `Component.js`

Author: done

> - for each page configuration, the configured component is added as a dependency to the template app module

Reviewer: ```suggestion
- For each page configuration, the configured component is added as a dependency to the template app module,
```

Author: done

> - if the page configuration contains a templateName, a dependency to that template view is added to the app

Reviewer: ```suggestion
- if the page configuration contains a `templateName`, a dependency to that template view is added to the app,
```

> - Otherwise, the class definition of the component is analyzed to find a default template view name. if found, a dependency to that view is added to the app module.

Reviewer 2: --> ". If found,"

> - For each page configuration, the configured component is added as a dependency to the template app module,
> - if the page configuration contains a `templateName`, a dependency to that template view is added to the app,
> - otherwise, the class definition of the component is analyzed to find a default template view name. if found, a dependency to that view is added to the app module.

Reviewer: ```suggestion
- For each page configuration, the configured component is added as a dependency to the template app module.
- If the page configuration contains a `templateName`, a dependency to that template view is added to the app.
- Otherwise, the class definition of the component is analyzed to find a default template view name. if found, a dependency to that view is added to the app module.
```

> - otherwise, the class definition of the component is analyzed to find a default template view name. if found, a dependency to that view is added to the app module

Reviewer: ```suggestion
- otherwise, the class definition of the component is analyzed to find a default template view name. if found, a dependency to that view is added to the app module.
```

The template component is analyzed in the following way:

> - precondition: template component class is defined in an AMD-style module, using define or sap.ui.define

Reviewer: ```suggestion
- precondition: the template component class is defined in an AMD-style module, using define or sap.ui.define
```

> - precondition: the module 'sap/suite/ui/generic/template/lib/TemplateAssembler' is imported

Reviewer 2: backticks for 'sap/suite/ui/generic/template/lib/TemplateAssembler'

Author: Done

> - precondition: a call to TemplateAssembler.getTemplateComponent is used to define the component class

Reviewer 2: backticks for TemplateAssembler.getTemplateComponent

Author: done

> - precondition: that call is used in a top level return statement of the factory function

Reviewer: ```suggestion
- precondition: that call is used in a top-level return statement of the factory function
```

- precondition: necessary parameters to that call are given as an object literal (no further coding)

> - precondition: the settings define a managed property 'metadata.properties.templateName' with a defaultValue of type string

Reviewer 2: backticks for 'metadata.properties.templateName', also for defaultValue

> - precondition: the settings define a managed property property 'metadata.properties.templateName' with a defaultValue of type string

Reviewer: ```suggestion
- precondition: the settings define a managed property 'metadata.properties.templateName' with a defaultValue of type string
```

The default value of the property represents the template view of the template component.

> The manifest of the template app in theory could specify an alternative template in

Reviewer: ```suggestion
The manifest of the template app could in theory specify an alternative template in
```

> component.settings.templateName.

Reviewer 2: backticks for component.settings.templateName

Author: done

### XML Template Analyzer

> The [XML Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLTemplateAnalyzer.js) tackles XMLViews and XMLFragments. It parses the XML, collects controls, and adds them as dependency to the ModuleInfo object.

Reviewer 2: --> "tackles `XMLView` and `XMLFragment`"; backticks for ModuleInfo

Author: done

> The [XML Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLTemplateAnalyzer.js) tackles XMLViews and XMLFragments. It parses the XML, collects controls and adds them as dependency to the ModuleInfo object.

Reviewer: ```suggestion
The [XML Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLTemplateAnalyzer.js) tackles XMLViews and XMLFragments. It parses the XML, collects controls, and adds them as dependency to the ModuleInfo object.
```

Additionally, some special dependencies are handled:

> - controller of the view

Reviewer: ```suggestion
- Controller of the view,
```

> - resource bundle (note: as locale dependent dependencies can't be modelled yet in the ModuleInfo, only a dependency to the development version (aka raw language) of the bundle will be added)

Reviewer: ```suggestion
- resource bundle (note: as locale-dependent dependencies can't be modelled yet in the ModuleInfo, only a dependency to the development version (aka raw language) of the bundle will be added),
```

> - component referenced via ComponentContainer control

Reviewer: ```suggestion
- component referenced via ComponentContainer control,
```

> - Controller of the view,
> - resource bundle (note: as locale-dependent dependencies can't be modelled yet in the ModuleInfo, only a dependency to the development version (aka raw language) of the bundle will be added),
> - component referenced via ComponentContainer control,
> - embedded fragments or views.

Reviewer: ```suggestion
- controller of the view
- resource bundle (note: as locale-dependent dependencies can't be modelled yet in the ModuleInfo, only a dependency to the development version (aka raw language) of the bundle will be added)
- component referenced via the ComponentContainer control
- embedded fragments or views
```

> - embedded fragments or views

Reviewer: ```suggestion
- embedded fragments or views.
```

> In an XMLView/XMLFragment, usually 3 categories of element nodes exist: Controls, aggregations, and non-UI5 nodes (e.g. XHTML or SVG).

Reviewer 2: --> "three categories"

Author: Done

> In an XMLView/XMLFragment, there usually exist 3 categories of element nodes: controls, aggregations and non-UI5 nodes (e.g. XHTML or SVG).

Reviewer: ```suggestion
In an XMLView/XMLFragment, usually 3 categories of element nodes exist: Controls, aggregations, and non-UI5 nodes (e.g. XHTML or SVG).
```

### XML Composite Analyzer

> The **XMLComposite** control is deprecated since UI5 Version 1.88. Nevertheless, UI5 Tooling will attempt to analyze the declaration of any such controls in a project. The [XML Composite Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLCompositeAnalyzer.js) searches for the name of the configured fragment containing the **XMLComposite** control.

Reviewer 2: --> "since version UI5 1.88" 
(as this makes searches more consistent)

Author: done in all occurrences

> The **XMLComposite** control is deprecated since UI5 Version 1.88. Nevertheless UI5 Tooling will attempt to analyze the declaration of any such controls in a project. The [XML Composite Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLCompositeAnalyzer.js) searches for the name of the configured fragment containing the **XMLComposite** control.

Reviewer: ```suggestion
The **XMLComposite** control is deprecated since UI5 Version 1.88. Nevertheless, UI5 Tooling will attempt to analyze the declaration of any such controls in a project. The [XML Composite Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLCompositeAnalyzer.js) searches for the name of the configured fragment containing the **XMLComposite** control.
```

> === "Name of the XMLComposite is equal with fragment name"

Reviewer: ```suggestion
=== "Name of the XMLComposite is equal to fragment name"
```

    ```javascript hl_lines="4"
    sap.ui.define([
        "sap/ui/core/XMLComposite"
    ], function(XMLComposite) {
        return XMLComposite.extend("composites.MyComposite", {} 
    });
    ```

=== "Dedicated fragment name"

    ```javascript hl_lines="5"
    sap.ui.define([
        "sap/ui/core/XMLComposite"
    ], function(XMLComposite) {
        return XMLComposite.extend("composites.MyComposite", {
            fragment: "composites.custom.MyComposite"
        } 
    });
    ```

## Library Initialization

> The [library.js Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/analyzeLibraryJS.js) checks every `library.js` file in the namespace of a library for occurences of a `sap/ui/core/Core#initLibrary` call. If so, the following information will be placed in the generated manifest.json:

Reviewer: ```suggestion
The [library.js Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/analyzeLibraryJS.js) checks every `library.js` file in the namespace of a library for occurences of a `sap/ui/core/Core#initLibrary` call. If found, the following information will be placed in the generated manifest.json:
```

- noLibraryCSS: false when the noLibraryCSS property had been set in the initLibrary info object
- types: string array with the names of the types contained in the library
- controls: string array with the names of the controls defined in the library
- elements: string array with the names of the elements defined in the library
- interfaces: string array with the names of the interfaces defined in the library

> When using `sap/ui/core/Core#initLibrary` requires the usage of "Simple Literals" for the parameter passed to this function call.

Reviewer: ```suggestion
`sap/ui/core/Core#initLibrary` requires the usage of "simple literals" for the parameters passed to this function call.
```

> Note: currently only the usage via the global `sap.ui.getCore().initLibrary` is supported by this analyzer. Requiring the `sap/ui/core/Core` and then call `Core.initLibrary` is not recognized by this analyzer.

Reviewer: ```suggestion
Note: Currently only the usage via the global `sap.ui.getCore().initLibrary` is supported by this analyzer. Requiring `sap/ui/core/Core` and then calling `Core.initLibrary` is not recognized by this analyzer.
```

## JSDoc

> The UI5 Tooling offers a JSDoc build, which is enhanced by UI5 specific JSDoc features.

Reviewer: ```suggestion
The UI5 Tooling offers a JSDoc build, which is enhanced by UI5-specific JSDoc features.
```

> An UI5 SDK can be build locally. To get more insights about the local UI5 SDK build setup, have a look at the [Developer Guide](https://github.com/SAP/openui5/blob/master/docs/developing.md#building-the-openui5-sdk-demo-kit).

Reviewer: ```suggestion
An UI5 SDK can be built locally. To get more insight into the local UI5 SDK build setup, have a look at our [Developer's Guide](https://github.com/SAP/openui5/blob/master/docs/developing.md#building-the-openui5-sdk-demo-kit).
```

> Currently, the resources needed for a UI5 SDK build are stored in [openui5](https://github.com/SAP/openui5/tree/master/lib/jsdoc) and in [ui5-builder](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc). This double maintenance is needed because these files are not part of the `sap.ui.core` library artefact, so building JSDoc for any library has no access to the needed resources. Therefore it is necessary to have these resources also available in the *ui5-builder*. This might change in future.

Reviewer: ```suggestion
Currently, the resources needed for a UI5 SDK build are stored in [openui5](https://github.com/SAP/openui5/tree/master/lib/jsdoc) and in [ui5-builder](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc). This double maintenance is needed because these files are not part of the `sap.ui.core` library artefact, so building JSDoc for any library has no access to the needed resources. It's therefore necessary to have these resources also available in the *ui5-builder*. This might change in the future.
```

> Following artefacts contributing to the JSDoc build:

Reviewer: ```suggestion
The following artefacts contribute to the JSDoc build:
```

- [jsdocGenerator.js](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/jsdocGenerator.js):
  Executes the actual JSDoc build. UI5 Tooling wrapper for `plugin.cjs`.
- [sdkTransformer.js](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/sdkTransformer.js):
  UI5 Tooling wrapper for `transformApiJson.cjs`.
- [apiIndexGenerator.js](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/apiIndexGenerator.js):
  UI5 Tooling wrapper for `createIndexFiles.cjs`.
- [createIndexFiles.cjs](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/lib/createIndexFiles.cjs):
  Creates API index resources from all `api.json` resources and by interpreting the `sap-ui-version.json` available in the given test resources directory. The resulting index resources (e.g. `api-index.json`,  `api-index-deprecated.json`,
  `api-index-experimental.json` and `api-index-since.json`) are only to be used in a UI5 SDK.
- [transformApiJson.cjs](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/lib/transformApiJson.cjs): 

>   Preprocesses `api.json` files for use in a UI5 SDKs. Transforms the `api.json` as created by the JSDoc build into a pre-processed `api.json` file suitable for the SDK. The pre-processing includes formatting of type references, rewriting of links and other time consuming calculations.

Reviewer: ```suggestion
  Preprocesses `api.json` files for use in UI5 SDKs. Transforms the `api.json` as created by the JSDoc build into a pre-processed `api.json` file suitable for the SDK. The pre-processing includes formatting of type references, rewriting of links, and other time-consuming calculations.
```

- [plugin.cjs](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/lib/ui5/plugin.cjs):

>   UI5 plugin for JSDoc3. The plugin adds the following UI5 specific tag definitions to JSDoc3.

Reviewer: ```suggestion
  UI5 plugin for JSDoc3. The plugin adds the following UI5-specific tag definitions to JSDoc3:
```

    - disclaimer
    - experimental
    - final
    - interface
    - implements

> 	- ui5-restricted and more

Reviewer: ```suggestion
	- ui5-restricted, and more
```

>     It furthermore listens to the following JSDoc3 events to implement additional functionality

Reviewer: ```suggestion
    It furthermore listens to the following JSDoc3 events to implement additional functionality:
```

>     - parseBegin: to create short names for all file that are to be parsed

Reviewer: ```suggestion
    - parseBegin: to create short names for all files that are to be parsed
```

>     - fileBegin: to write some line to the log (kind of a progress indicator)

Reviewer: ```suggestion
    - fileBegin: to write to the log (a kind of progress indicator)
```

>     - jsdocCommentFound: to pre-process comments, empty lines are used as paragraph markers a default visibility is added, legacy tag combinations used in JSdoc2 are converted to JSDoc3 conventions

Reviewer: ```suggestion
    - jsdocCommentFound: to pre-process comments, empty lines are used as paragraph markers, a default visibility is added, and legacy tag combinations used in JSDoc2 are converted to JSDoc3 conventions
```

    - newDoclet
    - parseComplete: remove undocumented/ignored/private doclets or duplicate doclets

>     Last but not least, it implements an astNodeVisitor to detect UI5 specific "extend" calls and to create documentation for the properties, aggregations etc. that are created with the "extend" call.

Reviewer: ```suggestion
    Last but not least, it implements an astNodeVisitor to detect UI5-specific "extend" calls and to create documentation for the properties, aggregations, etc. that are created with the "extend" call.
```

- [publish.cjs](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/lib/ui5/template/publish.cjs): JSDoc3 template for UI5 documentation generation.

    - adds missing namespaces

>     - determines the exports names of exported APIs

Reviewer: ```suggestion
    - determines the export names of exported APIs
```

    - writes out the `api.json` from the collected JSDoc information
	- calculates the inheritance hierarchy
	- checks for cyclic dependencies
	- removes unnecessary whitespace from an HTML document

> - [versionUtil.cjs](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/lib/ui5/template/utils/versionUtil.cjs): Provides helper methods to determine version related information.

Reviewer: ```suggestion
- [versionUtil.cjs](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc/lib/ui5/template/utils/versionUtil.cjs): Provides helper methods to determine version-related information.
```

# Code Analysis

During the build process, UI5 Tooling executes a static code analysis of your project. In the following sections the term "simple literal" is used for the following language features:

- string literals
- numeric literals

> - Boolean literals

Reviewer 3: Boolean is still upper case?

Reviewer: It's uppercase in the [SAP API Style Guide](https://help.sap.com/docs/SNG_4_APIDOC/53e39c8b7c924c28a2575be50bc09786/259573468585469f97651cfdfe0e4672.html), in SAPterm / [Terminology Search App](https://lpxtest-approuter-term.cfapps.sap.hana.ondemand.com/ui/), and in our very own internal [UA guidelines for standard formulations in the API Reference](https://workzone.one.int.sap/site#workzone-home&/wiki/show/QXEPJkMTXgPzsAcuGP8pYb).

Reviewer 3: but I see no reason why boolean would be different from string or numeric?

Reviewer 3: Oh, I guess it's because Boolean is named after George Boole, so it's rated as a name. From a technical point of view, I can't understand this. The term is long established and no longer associated with a person. JavaScript and TypeScript both use the lower case name, so for me this is the closest association in our context. 

But okay, we should stick to the guides.

Maybe type string is also named after [George String](https://www.discogs.com/de/artist/1613301-Georg-String) 🤷‍♂️

Reviewer: Because it's derived from a surname of a person - the mathematician George Boole. See, for example, 'Newtonian mechanics', 'Hamiltonian operator', 'Euclidian geometry', ... all these are uppercase as well.

- `null` / `undefined`

> - Template Literal

Author: Done.

- Arrays of the previous, spread operators will be ignored (not taken into account)
- Object Literals with simple literals as keys and values (again, spread operators will be ignored / not taken into account)

## Dependency Analysis

UI5 Tooling extracts dependency information from a project's code as outlined in the following chapters.
The APIs described in the following sections requiring the usage of "Simple Literals" when declaring dependencies.

### JSModule Analyzer

> The [JSModule Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/JSModuleAnalyzer.js) uses the ["Abstract Syntax Tree"](https://en.wikipedia.org/wiki/Abstract_syntax_tree) (AST) of a JavaScript file to decide whether a code block is executed *conditionally* or *unconditionally*.

Reviewer 3: I would swap the order of the paragraphs (3, 1, 2 instead of 1,2,3). It is more important to know that certain APIs are detected to find dependencies. Only then one can IMO understand that the found dependencies are classified as standard, conditional or dynamic.

Author: Done

> Besides this information, which is inherent to the JavaScript language, the analyzer uses additional knowledge about special API. For example, the factory function of an AMD-module is known to be executed when the module is executed, an IIFE is known to be executed immediately etc.

Author: done. Thx @codeworrior

- sap.ui.preload (restricted)
- sap.ui.require.preload (restricted)

The [JSModule Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/JSModuleAnalyzer.js) uses the ["Abstract Syntax Tree"](https://en.wikipedia.org/wiki/Abstract_syntax_tree) (AST) of a JavaScript file to decide whether a code block is executed *conditionally* or *unconditionally*.

The analyzer uses a set of rules to decide whether one of the above APIs is called whenever the module is executed or whether the API is only called under certain conditions.

For example, top-level code is always executed. Flow-control statements in JavaScript imply that certain blocks of code are only executed under certain conditions (for example, `if` blocks, `else` blocks, ...). Besides these inherent JavaScript rules, further common patterns are known to the analyzer, e.g. immediately invoked function expressions or the factor function of AMD modules.

Any dependencies found that - according to these rules - are always executed, are collected as eager (or standard) dependencies. Dependencies that are found on a code path that depends on certain conditions are collected as conditional dependencies.

> The bundling implemented by UI5 Tooling can either follow only eager dependencies (resolve:`true`) or additionally conditional dependencies (resolveConditional). For more information, see [Custom Bundling](https://sap.github.io/ui5-tooling/v3/pages/Configuration/#custom-bundling).

Reviewer 3: sholdn't config properties (resolve, resolveConditional) be wrapped in backticks?

Author: done

- any component dependency is added as a dependency to the Component.js module of that component. If the Component dependency is modeled as 'lazy', the module dependency will be added as 'conditional'

> - for each configured UI5 module, for which a type is configured, a module dependency to that type is added

Author: so "for each configured UI5 model, for which a type is configured, a module dependency to that type is added"?

> - for each configured UI5 module, for which a type is configured, a module dependency to that type is added

Reviewer 3: The proposed sentence sounds okay.

> - for each route that contains a view name, a module dependency to that view will be added

Author: done

> - for each route that contains a view name, a module dependency to that view will be added

Author: done

- precondition: the settings define a managed property property 'metadata.properties.templateName' with a
                defaultValue of type string
The default value of the property represents the template view of the template component.
The manifest of the template app in theory could specify an alternative template as setting.templateName,
but as of June 2017, this possibility is currently not used.

### Smart Template Analyzer

The [Smart Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/SmartTemplateAnalyzer.js) analyzes a Smart Template app and its underlying template components to collect dependency information. It searches for a `manifest.json` in the same folder. If it is found and if it is a valid JSON, an "sap.ui.generic.app" section is searched and evaluated in the following way

- for each page configuration, the configured component is added as a dependency to the template app module
- If the page configuration contains a templateName, a dependency to that template view is added to the app
- Otherwise, the class definition of the component is analyzed to find a default template view name

>    If found, a dependency to that view is added to the app module

Author: done

>    If found, a dependency to that view is added to the app module

Author: done

- precondition: template component class is defined in an AMD-style module, using define or sap.ui.define
- precondition: the module 'sap/suite/ui/generic/template/lib/TemplateAssembler' is imported
- precondition: a call to TemplateAssembler.getTemplateComponent is used to define the component class
- precondition: that call is used in a top level return statement of the factory function
- precondition: necessary parameters to that call are given as an object literal (no further coding)
- precondition: the settings define a managed property property 'metadata.properties.templateName' with a defaultValue of type string
The default value of the property represents the template view of the template component.
The manifest of the template app in theory could specify an alternative template in
component.settings.templateName.

### XML Template Analyzer

The [XML Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLTemplateAnalyzer.js) tackles XMLViews and XMLFragments. It parses the XML, collects controls and adds them as dependency to the ModuleInfo object.
Additionally, some special dependencies are handled:

- controller of the view

> - resource bundle (note: locale dependent dependencies can't be modeled yet in ModuleInfo)

Author: no its the locale but i have added "`sap/ui/core/Locale.js` dependent" to highlight this.

- component referenced via ComponentContainer control
- embedded fragments or views

In an XMLView/XMLFragment, there usually exist 3 categories of element nodes: controls, aggregations
of cardinality 'multiple' and non-UI5 nodes (e.g. XHTML or SVG). The third category usually
can be identified by its namespace. To distinguish between the first and the second
category, this analyzer uses a ResourcePool (provided by the caller and usually derived from the
library classpath). When the qualified node name is contained in the pool, it is assumed to
represent a control, otherwise it is ignored.

> In an XMLView/XMLFragment, aggregations of cardinality 0..1 (single) also appear as element nodes in the XML, they just must not contain more than one element.

Author: Thats true but we wanna highlight exactly this.

> In an XMLView/XMLFragment, usually three categories of element nodes exist: Controls, aggregations, and non-UI5 nodes (e.g. XHTML or SVG).

Reviewer 3: That paragraph is now a bit lost. Maybe my comment regarding the pool was misunderstood. I didn't want to suggest that the explanation of the lookup in general should be removed. My complaint was only about the (IMO implementation detail) that a ResourcePool is used for that.

Author: Done and reviewed by UA.

to implement the exactly same logic as used in the runtime XMLTemplateProcessor would be to
expensive and require too much runtime.

### XML Composite Analyzer

The **XMLComposite** control is deprecated since UI5 Version 1.88. Nevertheless UI5 Tooling will attempt to analyze the declaration of any such controls in a project. The [XML Composite Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLCompositeAnalyzer.js) searches for the name of the configured fragment containing the **XMLComposite** control.

=== "Name of the XMLComposite is equal with fragment name"

    ```javascript hl_lines="4"
    sap.ui.define([
        "sap/ui/core/XMLComposite"
    ], function(XMLComposite) {
        return XMLComposite.extend("composites.MyComposite", {} 
    });
    ```

=== "Dedicated fragment name"

    ```javascript hl_lines="5"
    sap.ui.define([
        "sap/ui/core/XMLComposite"
    ], function(XMLComposite) {
        return XMLComposite.extend("composites.MyComposite", {
            fragment: "composites.custom.MyComposite"
        } 
    });
    ```

## Library Initialization

> The [library.js Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/analyzeLibraryJS.js) checks every JavaScript file for occurences of a `sap/ui/core/Core#initLibrary` call. If so, the following information will be placed in the generated manifest.json:

Author: What about "Note: currently only the usage via the global `sap.ui.getCore().initLibrary` is supported by this analyzer. Requiring the `sap/ui/core/Core` and then call `Core.initLibrary` is not catched by this analyzer."

> The [library.js Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/analyzeLibraryJS.js) checks every JavaScript file for occurences of a `sap/ui/core/Core#initLibrary` call. If so, the following information will be placed in the generated manifest.json:

Reviewer 3: sounds good, maybe "understood" or "recognised" rather than "catched" (btw, I think it's "caught")

> The [library.js Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/analyzeLibraryJS.js) checks every JavaScript file for occurences of a `sap/ui/core/Core#initLibrary` call. If so, the following information will be placed in the generated manifest.json:

Author: updated to "recognized" --> AE is preferred when I remember correctly

- noLibraryCSS: false when the noLibraryCSS property had been set in the initLibrary info object
- types: string array with the names of the types contained in the library
- controls: string array with the names of the controls defined in the library
- elements: string array with the names of the elements defined in the library
- interfaces: string array with the names of the interfaces defined in the library

When using `sap/ui/core/Core#initLibrary` requires the usage of "Simple Literals" for the parameter passed to this function call.

## JSDoc

The UI5 Tooling offers a JSDoc build, which is enhanced by UI5 specific JSDoc features.

An UI5 SDK can be build locally. To get more insights about the local UI5 SDK build setup, have a look at the [Developer Guide](https://github.com/SAP/openui5/blob/master/docs/developing.md#building-the-openui5-sdk-demo-kit).

> Currently, the resources needed for a UI5 SDK build are stored in [openui5](https://github.com/SAP/openui5/tree/master/lib/jsdoc) and in [ui5-builder](https://github.com/SAP/ui5-builder/blob/main/lib/processors/jsdoc). This double maintenance is needed because these files are not part of the `sap.ui.core` library artefact, so building JSDoc for any library has no access to the needed resources. Therefore it is necessary to have these resources also available in the *ui5-builder*. This might change in future.

Author: Thats true, but I guess many content of this site is only of interest for us. but when we write it down its written down :-)

- precondition: template component class is defined in an AMD-style module, using define or sap.ui.define
- precondition: the module 'sap/suite/ui/generic/template/lib/TemplateAssembler' is imported
- precondition: a call to TemplateAssembler.getTemplateComponent is used to define the component class
- precondition: that call is used in a top level return statement of the factory function
- precondition: necessary parameters to that call are given as an object literal (no further coding)
- precondition: the settings define a managed property property 'metadata.properties.templateName' with a defaultValue of type string
The default value of the property represents the template view of the template component.
The manifest of the template app in theory could specify an alternative template in
component.settings.templateName.

### XML Template Analyzer

The [XML Template Analyzer](https://github.com/SAP/ui5-builder/blob/main/lib/lbt/analyzer/XMLTemplateAnalyzer.js) tackles XMLViews and XMLFragments. It parses the XML, collects controls and adds them as dependency to the ModuleInfo object.
Additionally, some special dependencies are handled:

- controller of the view

> - resource bundle (note: `sap/ui/core/Locale.js` dependent dependencies can't be modeled yet in ModuleInfo)

Reviewer 3: I wouldn't mention sap/ui/core/Locale here. The concept of language dependent files is not bound to the runtime class that represents the languages/locales. 
Proposal:
> (note: as locale dependent dependencies can't be modelled yet in the ModuleInfo, only a dependency to the development version (aka raw language) of the bundle will be added)

Author: done

- component referenced via ComponentContainer control
- embedded fragments or views

In an XMLView/XMLFragment, there usually exist 3 categories of element nodes: controls, aggregations

> of cardinality 'multiple' and non-UI5 nodes (e.g. XHTML or SVG). The third category usually

Reviewer 3: as discussed earlier, also aggregations of cardinality 0..1 can appear in the XML. Therefore, I would omit the part about the cardinality.

Author: completely removed the content when talking about cardinalities

can be identified by its namespace. To distinguish between the first and the second
category, this analyzer uses a ResourcePool (provided by the caller and usually derived from the

> library classpath). When the qualified node name is contained in the pool, it is assumed to

Reviewer 3: Again, I would not mention implementation details in the user documentation. From a user's perspective, it's rather a "check against the resources in a project, including its dependencies" (Non-Java developers might not be familiar with the term "classpath")

Author: completely removed the content when talking about cardinalities

represent a control, otherwise it is ignored.

In an XMLView/XMLFragment, aggregations of cardinality 0..1 (single) also appear as element nodes in the XML, they just must not contain more than one element.

> In certain cases this might give wrong results, but loading the metadata for each control

Reviewer 3: "In certain cases this ..."

The context of "this" is not clear as there's now a paragraph between the "this" (usage of pool to determine whether an XML node is a control or an aggregation) and the sentence here.

I would either move the aggregation part elsewhere (maybe I would even completely omit the talk about cardinalities) or a short form of "this" has to be repeated again here, .e.g. "this check against the set of known controls..."

Author: completely removed the content when talking about cardinalities
